device/transport/mqtt/src/mqtt_twin_receiver.ts

```diff
@@ -47,80 +50,122 @@ export class MqttTwinReceiver extends EventEmitter {
-        process.nextTick(() => {
-          self._mqtt.subscribe(postTopic, { qos: 0 }, (err, transportObject) => {
+    });
+
+    this.on('removeListener', (eventName) => {
+      if (eventName === MqttTwinReceiver.responseEvent || eventName === MqttTwinReceiver.postEvent) {
+        if (this.listeners('response').length + this.listeners('post').length <= 0) {
+          this._fsm.handle('unsubscribe', (err) => {
             if (err) {
-              self._handleError(err);
-            } else {
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_025: [** If the `subscribed` event is subscribed to, a `subscribed` event shall be emitted after an MQTT topic is subscribed to. **]** */
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_028: [** When the `subscribed` event is emitted, the parameter shall be an object which contains an `eventName` field and an `transportObject` field. **]** */
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_027: [** When the `subscribed` event is emitted because the post MQTT topic was subscribed, the `eventName` field shall be the string 'post' **]** */
-              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_029: [** When the subscribed event is emitted, the `transportObject` field shall contain the object returned by the library in the subscription response. **]** */
-              self.emit(MqttTwinReceiver.subscribedEvent, { 'eventName' : MqttTwinReceiver.postEvent, 'transportObject' : transportObject });
+              debug('error while unsubscribing: ' + err.toString());
+              /* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_024: [** When the `error` event is emitted, the first parameter shall be an error object obtained via the MQTT `translateError` module. **]** */
+              this.emit(MqttTwinReceiver.errorEvent, translateError(err));
             }
           });
-        });
+        }
       }
-    }
+    });
+
+    this._fsm = new machina.Fsm({
+      namespace: 'device-mqtt',
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: (callback, err) => {
+            if (callback) {
+              callback(err);
+            } else if (err) {
+              this.emit('error', translateError(err));
+            }
+          },
+          subscribe: (callback) => this._fsm.transition('subscribing', callback),
+          unsubscribe: (callback) => callback()
+        },
+        subscribing: {
+          _onEnter: (callback) => {
+            debug('subscribing to twin topics');
+            this._mqtt.on('message', this._onMqttMessage.bind(this));
+            this._mqtt.subscribe(responseTopic, { qos: 0 }, (err, responseTopicResult) => {
+              if (err) {
+                this._fsm.transition('unsubscribed', callback, err);
+              } else {
+                debug('subscribed to response topic');
+                this._mqtt.subscribe(postTopic, { qos: 0 }, (err, postTopicResult) => {
```

Comment:
>subscribe( [](start = 27, length = 10)

If this fails but the response topic subscription succeeds you'll pop right to unsubscribed with an outstanding subscription to the response.  Is this ok?

Reply:
probably not...

---
In reply to: [145750640](https://github.com/Azure/azure-iot-sdk-node/pull/111#discussion_r145750640) [](ancestors = 145750640)